ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -692,64 +664,135 @@ private JoinNode buildJoin(final Join root, final String prefix, final boolean i
                 "Invalid join condition:"
                     + " table-table joins require to join on the primary key of the right input"
                     + " table. Got %s = %s.",
-                joinInfo.hasFlippedJoinCondition() ? rightExpression : leftExpression,
-                joinInfo.hasFlippedJoinCondition() ? leftExpression : rightExpression
+                joinInfo.getFlippedLeftJoinExpression(),
+                joinInfo.getFlippedRightJoinExpression()
            ));
       }
 
       if (joinOnNonKeyAttribute(leftExpression, leftNode)) {
         // foreign key join detected
 
-        if (ksqlConfig.getBoolean(KsqlConfig.KSQL_FOREIGN_KEY_JOINS_ENABLED)) {
+        return verifyForeignKeyJoin(joinInfo, leftNode, rightNode);
+      } else {
+        // primary key join detected
+
+        final SqlType leftKeyType = Iterables.getOnlyElement(leftNode.getSchema().key()).type();
+        final SqlType rightKeyType = Iterables.getOnlyElement(rightNode.getSchema().key()).type();
+
+        verifyJoinConditionTypes(
+            leftKeyType,
+            rightKeyType,
+            leftExpression,
+            rightExpression,
+            joinInfo.hasFlippedJoinCondition()
+        );
+      }
+    }
 
-          if (joinInfo.getType().equals(JoinType.OUTER)) {
-            throw new KsqlException(String.format(
-                "Invalid join type:"
-                    + " full-outer join not supported for foreign-key table-table join."
-                    + " Got %s %s %s.",
-                joinInfo.getLeftSource().getDataSource().getName().text(),
-                joinType,
-                joinInfo.getRightSource().getDataSource().getName().text()
-            ));
-          }
-
-          // after we lift this n-way join restriction, we should be able to support FK-joins
-          // at any level in the join tree, even after we add right-deep/bushy join tree support,
-          // because a FK-join output table has the same PK as its left input table
-          if (!(leftNode instanceof DataSourceNode)
-              || !(rightNode instanceof DataSourceNode)) {
-            throw new KsqlException(String.format(
-                "Invalid join condition: foreign-key table-table joins are not "
-                    + "supported as part of n-way joins. Got %s = %s.",
-                joinInfo.hasFlippedJoinCondition() ? rightExpression : leftExpression,
-                joinInfo.hasFlippedJoinCondition() ? leftExpression : rightExpression
-            ));
-          }
+    return Optional.empty();
+  }
 
-          if (!(leftExpression instanceof ColumnReferenceExp)) {
-            throw new KsqlException(String.format(
-                "Invalid join condition: foreign-key table-table joins with expressions "
-                    + "are not supported. Got %s = %s."
-            ));
-          }
+  private static void verifyStreamTableJoin(final JoinInfo joinInfo, final PlanNode rightNode) {
```

Comment:
I'm assuming this diff does not include any logical change for stream-table join, just refactoring here. LMK if it contains any logical changes.

Also, I'm wondering why we can declare this function a static but cannot do the same for `verifyForeignKeyJoin`?

Reply:
Correct. I extracted those only to make checkstyle happy so we can remove the supressions ` CHECKSTYLE_RULES.OFF: CyclomaticComplexity|NPathComplexity`

`verifyForeignKeyJoin` accesses member variable `ksqlConfig` atm, so it cannot be `static.` If we remove the feature flag at some point in the future, we could make it `static`, too.